compiler/src/main/kotlin/ast/reduce.kt

```diff
@@ -150,47 +151,45 @@ private fun reduceStmt(node: ParseTree): Stmt {
     return when (node) {
         is CellmataParser.Assign_stmtContext -> reduceStmt(node.assignment())
         is CellmataParser.AssignmentContext -> AssignStmt(
-            ctx = node,
-            expr = reduceExpr(node.expr())
+            ctx = SourceContext(node),
+            expr = reduceExpr(node.expr()),
+            ident = node.var_ident().text, // TODO LHS Array, e.g.: "a[0] = 5;"
+            isDeclaration = node.STMT_LET() != null
         )
         is CellmataParser.If_stmtContext -> IfStmt(
-            ctx = node,
+            ctx = SourceContext(node),
             conditionals = listOf( // Create list of list of ConditionalBlocks, then flatten to list of ConditionalBlocks
                 listOf(
                     ConditionalBlock( // If clause
-                        ctx = node,
+                        ctx = SourceContext(node),
                         expr = reduceExpr(node.if_stmt_if().if_stmt_block().if_stmt_condition().expr()),
                         block = reduceCodeBlock(node.if_stmt_if().if_stmt_block().code_block())
                     )
                 ),
                 node.if_stmt_elif().map { // Elif clauses
                     ConditionalBlock(
-                        ctx = it,
+                        ctx = SourceContext(it),
                         expr = reduceExpr(it.if_stmt_block().if_stmt_condition().expr()),
                         block = reduceCodeBlock(it.if_stmt_block().code_block())
                     )
                 }
             ).flatten(),
-            elseBlock = when (node.if_stmt_else()) {
-                null -> null
-                is CellmataParser.If_stmt_elseContext -> reduceCodeBlock(node.if_stmt_else().code_block())
-                else -> throw AssertionError("Unexpected tree node")
-            }
+            elseBlock = if (node.if_stmt_else() == null) null else reduceCodeBlock(node.if_stmt_else().code_block())
         )
-        is CellmataParser.For_stmtContext -> ForStmt(
-            ctx = node,
-            initPart = AssignStmt(node.for_init().assignment(), expr = reduceExpr(node.for_init().assignment().expr())),
+        is CellmataParser.For_stmtContext -> ForLoopStmt(
+            ctx = SourceContext(node),
+            initPart = reduceStmt(node.for_init()) as AssignStmt,
             condition = reduceExpr(node.for_condition().expr()),
-            postIterationPart = AssignStmt(node.for_post_iteration().assignment(), expr = reduceExpr(node.for_post_iteration().assignment().expr())),
+            postIterationPart = reduceStmt(node.for_post_iteration()) as AssignStmt,
```

Comment:
The cast to `AssignStmt` will cause a crash if `ErrorStmt` is returned

Reply:
The statement can only be an `AssignStmt` due to the grammar rules. If the statement is an `ErrorStmt` the error happened somewhere else, and we should already have raised an error.